plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,501 @@
+unsafeInferKind ctx ty =
+  case inferKind ctx ty of
+    Nothing -> error "inferKind"
+    Just k  -> k
+
+-- | Shrink a type in a context assuming that it is of kind *.
+shrinkType :: HasCallStack
+           => Map TyName (Kind ())
+           -> Type TyName DefaultUni ()
+           -> [Type TyName DefaultUni ()]
+shrinkType ctx ty = map snd $ shrinkKindAndType ctx (Star, ty)
+
+-- | Shrink a type of a given kind in a given context in a way that keeps its kind
+shrinkTypeAtKind :: HasCallStack
+                 => Map TyName (Kind ())
+                 -> Kind ()
+                 -> Type TyName DefaultUni ()
+                 -> [Type TyName DefaultUni ()]
+shrinkTypeAtKind ctx k ty = [ ty' | (k', ty') <- shrinkKindAndType ctx (k, ty), k == k' ]
+
+-- | Check well-kindedness of a type in a context
+checkKind :: Map TyName (Kind ()) -> Type TyName DefaultUni () -> Kind () -> Bool
+checkKind ctx ty k = case ty of
+  TyVar _ x        -> Just k == Map.lookup x ctx
+  TyFun _ a b      -> k == Star && checkKind ctx a k && checkKind ctx b k
+  TyApp _ a b | Just kb <- inferKind ctx b -> checkKind ctx a (KindArrow () kb k) && checkKind ctx b kb
+              | otherwise                  -> False
+  TyLam _ x kx b
+    | KindArrow _ ka kb <- k -> kx == ka && checkKind (Map.insert x kx ctx) b kb
+    | otherwise              -> False
+  TyForall _ x kx b -> k == Star && checkKind (Map.insert x kx ctx) b k
+  TyBuiltin _ b    -> k == builtinKind b
+  TyIFix{}         -> error "checkKind: TyIFix"
+
+-- | Generate an arbitrary kind and closed type of that kind.
+genKindAndType :: Gen (Kind (), Type TyName DefaultUni ())
+genKindAndType = do
+  k <- arbitrary
+  t <- genClosedType_ k
+  return (k, t)
+
+-- | Normalize a type, throw an error if normalization fails due to e.g. wellkindedness issues.
+normalizeTy :: Type TyName DefaultUni () -> Type TyName DefaultUni ()
+normalizeTy ty = case runQuoteT $ normalizeType ty of
+  Left _                 -> error "normalizeTy"
+  Right (Normalized ty') -> ty'
+
+-- | Generate a context of free type variables with kinds
+genCtx :: Gen (Map TyName (Kind ()))
+genCtx = do
+  let m = 20
+  n <- choose (0, m)
+  let allTheVarsCalledX = [ TyName $ Name (fromString $ "x" ++ show i) (toEnum i) | i <- [1..m] ]
+  shuf <- shuffle allTheVarsCalledX
+  let xs = take n shuf
+  ks <- vectorOf n arbitrary
+  return $ Map.fromList $ zip xs ks
+
+-- CODE REVIEW: this should probably go somewhere else (??), where? Does it already exist?!
+instance Arbitrary (Kind ()) where
+  arbitrary = sized $ arb . (`div` 3)
+    where
+      arb 0 = pure $ Star
+      arb n = frequency [(4, pure $ Star),
+                         (1, (:->) <$> arb (div n 6) <*> arb (div (5 * n) 6))]
+  shrink Star      = []
+  shrink (a :-> b) = [b] ++ [a' :-> b' | (a', b') <- shrink (a, b)]
+    -- Note: `a` can have bigger arity than `a -> b` so don't shrink to it!
```

Comment:
It seems like the shrinking order relation is not just about arity? Here after `++` the arity doesn't change, so is it arity-first size-second or what? This definitely needs a Note (mentioned that already above).